nltk/twitter/util.py

```diff
@@ -0,0 +1,362 @@
+    Extract field values from a full tweet and return them as a list
+
+    :param json tweet: The tweet in JSON format
+    :param list fields: The fields to be extracted from the tweet
+    :rtype: list(str)
+    """
+    out = []
+    for field in fields:
+        try:
+            _add_field_to_out(tweet, field, out)
+        except TypeError:
+            raise RuntimeError('Fatal error when extracting fields. Cannot find field ', field)
+    return out
+
+
+def _add_field_to_out(json, field, out):
+    if isinstance(field, dict):
+        for key, value in field.items():
+            _add_field_to_out(json[key], value, out)
+    else:
+        if isinstance(field, compat.string_types):
+            out += [json[field]]
+        else:
+            out += [json[value] for value in field]
+
+def _get_entity_recursive(json, entity):
+    if not json:
+        return None
+    elif isinstance(json, dict):
+        for key, value in json.items():
+            if key == entity:
+                return value
+            '''
+            'entities' and 'extended_entities' are wrappers in twitter json
+            structure that contain other twitter objects. See:
+            https://dev.twitter.com/overview/api/entities-in-twitter-objects
+            '''
+            if key == 'entities' or key == 'extended_entities':
+                candidate = _get_entity_recursive(value, entity)
+                if candidate != None:
+                    return candidate
+        return None
+    elif isinstance(json, list):
+        for item in json:
+            candidate = _get_entity_recursive(item, entity)
+            if candidate != None:
+                return candidate
+        return None
+    else:
+        return None
+
+def json2csv(fp, outfile, fields, encoding='utf8', errors='replace'):
+    """
+    Extract selected fields from a file of line-separated JSON tweets and
+    write to a file in CSV format.
+
+    This utility function allows a file of full tweets to be easily converted
+    to a CSV file for easier processing. For example, just tweetIDs or
+    just the text content of the tweets can be extracted.
+
+    Additionally, the function allows combinations of fields of other Twitter
+    objects (mainly the users, see below).
+
+    For Twitter entities (e.g. hashtags of a tweet), and for geolocation, see
+    `json2csv_entities`
+
+    :param str infile: The name of the file containing full tweets
+
+    :param str outfile: The name of the text file where results should be\
+    written
+
+    :param list fields: The list of fields to be extracted. Useful examples\
+    are 'id_str' for the tweetID and 'text' for the text of the tweet. See\
+    <https://dev.twitter.com/overview/api/tweets> for a full list of fields.
+    e. g.: ['id_str'], ['id', 'text', 'favorite_count', 'retweet_count']
+    Addionally, it allows IDs from other Twitter objects, e. g.,\
+    ['id', 'text', {'user' : ['id', 'followers_count', 'friends_count']}]
```

Comment:
Using dict to specify nested objects leaves column order unspecified, which is problematic if headers are not output; and if headers are output the column names should be specified here. (Presumably one can use an `OrderedDict` to solve the first of those problems but who does?) Why not just use a notation like `'user.id', 'user.followers_count'` that keeps the parameter as flat as the output?